pkg/distributor/distributor.go

```diff
@@ -739,70 +740,111 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
 
 // Validates a single series from a write request.
 // May alter timeseries data in-place.
+// Returns a boolean stating if the timeseries should be removed from the request, and an error explaining the first validation finding.
 // The returned error may retain the series labels.
 // It uses the passed nowt time to observe the delay of sample timestamps.
-func (d *Distributor) validateSeries(nowt time.Time, ts *mimirpb.PreallocTimeseries, userID, group string, skipLabelValidation, skipLabelCountValidation bool, minExemplarTS, maxExemplarTS int64) error {
+func (d *Distributor) validateSeries(nowt time.Time, ts *mimirpb.PreallocTimeseries, userID, group string, skipLabelValidation, skipLabelCountValidation bool, minExemplarTS, maxExemplarTS int64) (bool, error) {
 	if err := validateLabels(d.sampleValidationMetrics, d.limits, userID, group, ts.Labels, skipLabelValidation, skipLabelCountValidation); err != nil {
-		return err
+		return true, err
 	}
 
 	now := model.TimeFromUnixNano(nowt.UnixNano())
 
-	for _, s := range ts.Samples {
+	timestamps := make(map[int64]struct{})
```

Comment:
Some suggestions:

1. To make this code a bit more readable, I would move the validation of a series samples and histograms to a dedicated function (e.g. `validateSeriesSamples()` and `validateSeriesHistograms()`). I don't feel strong about this, but I'm just noting `validateSeries()` function is getting quite big and a bit messy.
2. Most requests have 1 sample per series. To keep performance unchanged for the happy case, I would skip the duplicated samples check at all if there's only 1 series (so that we don't allocate `map[int64]struct{}` at all). Moving the logic to a dedicated function will make it easier to circuit break (e.g. `if len(rs.Samples) == 1 { return validateSample() }`)
3. In most cases, we don't expected duplicated samples for the same timestamp. For this reason, I would pre-allocate `make(map[int64]struct{})` to reduce memory re-allocations. We could pre-allocated to the `min(len(ts.Samples), 100))` where 100 is a reasonable max non-duplicate samples we expect in a single request)

Reply:
I have applied all your suggestions. The first one made the code a bit more complicated, but I tried to make it as much readable as possible.